### PR TITLE
[Perf][Fun-CosyVoice3] Batch HiFT vocoder inference

### PR DESCRIPTION
**Pull request by wirybeaver**
## Motivation

Fun-CosyVoice3's vocoder scheduler can coalesce requests, but HiFT was still invoked once per request. This implements the HiFT batch-inference item from the CosyVoice3 optimization roadmap without changing the existing Flow batching behavior.

## Modifications

- retain the existing Flow inference path and group its resulting mel tensors by exact frame length
- invoke HiFT once per compatible group without padding variable-length mel tensors
- restore waveforms to original request order and preserve each request's sample rate
- reject unexpected HiFT output batch cardinality
- add unit coverage for mixed-length grouping, batched tensor shapes, output ordering, and cardinality errors

## Related Issues

Related to #1652.

## Accuracy Test

- Fun-CosyVoice3 unit suite: 42 passed
- focused vocoder suite: 8 passed
- adjacent dots.tts and Fish Audio vocoder suites: 14 passed
- changed-line coverage: 100% (48/48)

The batching path only combines equal-length mel tensors; it does not pad variable-length inputs.

A paired real-checkpoint HiFT replay on H100 used three fixed speech-feature tensors extracted with the pinned CosyVoice frontend (207, 230, and 267 mel frames). Each batch repeated the same tensor so serial and batched calls consumed identical inputs. All output cardinality checks passed, all waveforms were finite, and batch-size-one output was exactly identical.

For batch sizes 2, 4, and 8, the observed numerical differences versus serial inference were input-dependent: RMS difference ranged from `3.88e-4` to `3.60e-3`, SNR from `26.38` to `37.94 dB`, and maximum absolute difference reached `0.0730`. At batch 8:

- 207 frames: RMS `0.003601`, SNR `26.38 dB`
- 230 frames: RMS `0.002045`, SNR `33.84 dB`
- 267 frames: RMS `0.000894`, SNR `30.69 dB`

These differences are larger than the earlier independent validation on another mel set (`3e-4` to `1.8e-3`, `41-57 dB`), indicating that the numerical delta depends on the input. No perceptual-equivalence claim is made from these metrics alone. The earlier real-checkpoint serving validation reported sane waveforms, correct ordering/sample rates, and no request failures: https://github.com/sgl-project/sglang-omni/pull/1655#issuecomment-5377558074.

### HiFT-only replay

The benchmark ran on one NVIDIA H100 80GB HBM3 with driver `595.71.05`, PyTorch `2.13.0+cu130`, CUDA `13.0`, bf16 autocast, and `FunAudioLLM/Fun-CosyVoice3-0.5B-2512`. Each cell used one warmup followed by three measured repeats of 100 iterations. The input fixture SHA-256 was `ced543960dda907d533446d9d79e1921be924adb77ce68333f37110839953409`.

The replay loads the same checkpoint HiFT object once and alternates the two call shapes across repeats:

- serial: invoke HiFT once per mel
- batched: concatenate equal-length mels and invoke HiFT once, then split outputs

At batch 8, mean GPU-only latency per item improved by about 3x:

- 207 frames: `45.030 ± 5.163 ms` → `15.457 ± 3.174 ms` (`2.91x`)
- 230 frames: `44.085 ± 5.892 ms` → `14.681 ± 0.742 ms` (`3.00x`)
- 267 frames: `44.919 ± 7.562 ms` → `15.102 ± 3.116 ms` (`2.97x`)

Including output splitting and CPU transfer, batch-8 latency per item improved by `3.03x`, `3.26x`, and `3.04x` for the same three lengths. GPU-only speedup was `1.94-1.98x` at batch 2, `2.34-2.80x` at batch 4, and `2.91-3.00x` at batch 8. Batch-size-one timing varied between `0.85x` and `1.07x` because of repeat/order noise, while its outputs remained exactly equal.

At batch 8, peak allocated memory was approximately `457/494/550 MiB` for 207/230/267 frames, versus `177/181/190 MiB` for serial calls. All 144 timing records were finite and positive, and all 12 accuracy records passed shape, cardinality, and finite-output validation.

This is a mechanism-level benchmark: the checkpoint-provided HiFT implementation is identical between Base and Head; the PR changes call orchestration. Measuring serial and batched calls against the same loaded object and identical mel inputs isolates that change. It intentionally excludes AR generation, Flow, HTTP, preprocessing, scheduler queueing, and mixed-length grouping. No new end-to-end replay was run as part of this validation; independent end-to-end results are linked above.

## End-to-end Benchmark (subsequent validation)

A subsequent end-to-end serving benchmark compared this PR against its exact base using the direct-container, prebuilt-application-image approach. Both application images were built from the same pinned CI image and differed only in the sglang-omni source revision:

- Base: `ef9479460420933d58c243d50befe18e6fe8ce9d`
- Head: `01f93554999c58823c3666233fa8c1fd56a1c635`
- Parent CI image: `hongccc/sglang-omni@sha256:02a85f00438c901c72a2eb2ef738974a807f63af3d13084445604f3344067b19`

The runs used one H100 NVL (`95,830 MiB`, driver `580.82.09`, CUDA `13.0.3`) and retained the same physical GPU UUID across all image activations. To mitigate temporal effects, complete outer blocks were interleaved `Base A -> Head A -> Head B -> Base B`, with a fresh managed server in every block.

The workload used `FunAudioLLM/Fun-CosyVoice3-0.5B-2512`, the pinned first 128 rows of the English SeedTTS split, voice cloning with flat references, non-streaming WAV, bf16/default precision, one warmup request, `max_running_requests=64`, and `cuda_graph_max_bs=64`. The 128-row prefix was used because an initial c=1 rate check showed that four full 1,088-row c=1 blocks alone would exceed the bounded benchmark budget. All 12 measured points completed `128/128` requests with zero failures.

Base and Head below are arithmetic means of their two outer blocks. Delta is `(Head / Base - 1) * 100`; positive is better for throughput and worse for latency/RTF.

| Conc. | Metric | Base | Head | Delta |
|---:|---|---:|---:|---:|
| 1 | Audio throughput (audio s/s) | 4.1245 | 4.1195 | -0.12% |
| 1 | Request throughput (req/s) | 0.9740 | 0.9735 | -0.05% |
| 1 | Mean latency (s) | 1.0265 | 1.0275 | +0.10% |
| 1 | P95 latency (s) | 1.5480 | 1.5610 | +0.84% |
| 1 | Mean RTF | 0.2612 | 0.2617 | +0.15% |
| 1 | P95 RTF | 0.4568 | 0.4582 | +0.28% |
| 8 | Audio throughput (audio s/s) | 15.8595 | 16.2740 | +2.61% |
| 8 | Request throughput (req/s) | 3.4525 | 3.5575 | +3.04% |
| 8 | Mean latency (s) | 2.2830 | 2.2265 | -2.47% |
| 8 | P95 latency (s) | 3.0710 | 3.1295 | +1.90% |
| 8 | Mean RTF | 0.5688 | 0.5556 | -2.31% |
| 8 | P95 RTF | 0.9086 | 0.8976 | -1.22% |
| 16 | Audio throughput (audio s/s) | 15.6205 | 17.8915 | +14.54% |
| 16 | Request throughput (req/s) | 3.4560 | 3.4115 | -1.29% |
| 16 | Mean latency (s) | 4.4835 | 4.5570 | +1.64% |
| 16 | P95 latency (s) | 5.7810 | 5.8450 | +1.11% |
| 16 | Mean RTF | 1.1240 | 1.1037 | -1.80% |
| 16 | P95 RTF | 1.8610 | 1.6966 | -8.83% |

The end-to-end result does **not** show a large request-level improvement. The c=1 control is unchanged. At c=8, aggregate changes are small and the paired-order audio-throughput deltas disagree (`-5.18%` and `+11.09%`). At c=16, audio throughput is `+14.54%`, but Head also generated about 16% more output tokens/audio duration; request throughput is `-1.29%`, mean latency is `+1.64%`, and P95 latency is `+1.11%`. Mean/P95 RTF improve by `1.80%`/`8.83%`. Thus, the mechanism-level HiFT speedup does not translate into a large end-to-end request-level gain for this tested workload.

Published benchmark images:

- Base: `rfgleamchaser/sglang-omni-benchmark@sha256:4726e3b81065dbe639d627a16cd2e248be3d3148321cb78c0bff0680bbd4bbd4`
- Head: `rfgleamchaser/sglang-omni-benchmark@sha256:a2e6e325d3d80bd0c4e9b9b257d13112f741830329c2a8309d649dbe733b5985`

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed. (No user-facing documentation change required.)
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

### CODE REVIEW

**Comment by Kozmosa** [comment]
hi, i have an H100 box with the Fun-CosyVoice3-0.5B-2512 checkpoint set up, and since the description mentions no CUDA GPU was available i ran the validation you were waiting on. hope it helps:

unit tests: `tests/unit_test/fun_cosyvoice3/` 42 passed on your branch.

serving (real checkpoint, 1x H100, 3 stages colocated): 24 mixed-length requests at concurrency 8 and 16, 0 errors, sane waveforms, per-request sample rate and result order all correct.

e2e throughput at ~3.5-5.4s utterances is within noise of the unbatched vocoder (conc 8: 3.28 vs 3.27 req/s, conc 16: 3.60 vs 3.52 req/s). AR + flow dominate the latency here so the batching does not move e2e numbers at this scale. the HiFT kernel itself does get faster though, measured in isolation on identical 200-token mels: 18.1 ms/item at batch 1, 12.3 at batch 4, 11.2 at batch 8+.

one small thing: `_token2wav` has no production callers after this change, only `test_vocoder.py` reaches into it directly. fine either way, just noticed.

File: sglang_omni/models/fun_cosyvoice3/stages.py [hunk 1]
```diff
@@ -176,8 +208,25 @@ def _token2wav(
...
             )
-        tts_speech, _ = self._hift.inference(speech_feat=tts_mel, finalize=True)
-        return tts_speech.detach().cpu()
+        return tts_mel
+
+    def _mel2wav_batch(self, mel_tensors: list[torch.Tensor]) -> list[torch.Tensor]:
+        speech_feat = (
+            mel_tensors[0] if len(mel_tensors) == 1 else torch.cat(mel_tensors, dim=0)
+        )
+        tts_speech, _ = self._hift.inference(
```
Comment (Kozmosa) [inline sglang_omni/models/fun_cosyvoice3/stages.py:603]: reusing the same mels and comparing batch N against batch 1 on the real HiFT, output is not bitwise identical: rms diff 3e-4 ~ 1.8e-3, snr 41-57 db, worst item 41.2 db at batch 8. sounds transparent to me, but maybe worth a line in the description since the no-padding grouping reads as "same output as unbatched" and it actually differs a bit through batch-dependent kernel numerics

Comment (wirybeaver) [reply sglang_omni/models/fun_cosyvoice3/stages.py:603]: Added the measured RMS/SNR range and the batch-dependent kernel-numerics caveat to the Accuracy Test section. Thanks for running this validation.

**Comment by PoTaTo-Mika** [comment]
Closed as finished. I've add you as a co-author in the new pull request for vocoder.

**Comment by wirybeaver** [comment]
superseded by https://github.com/sgl-project/sglang-omni/pull/1844
